[TASK] Update DragDrop.js

monosize opened the pull request:
Using the new DragDrop.js from TYPO3 8, which have changed from the jQuery UI sortable to droppable.
Copying the TYPO3 8 css for drag and drop.
Better drag and drop feeling.
Worked also in TYPO3 7.6 ;-)

The copy feature during drag and drop (holding the cmd-key) in TYPO3 8 is disabled in TYPO3 7.6.
This is done with a check if the localized text for 'dragdrop.copy.message' exists.

Fixed a z-index bug during dragging an element outside from a grid column.
Hide the dragged element in the column. This is useful for an element which contains a flux grid.

Tested with in TYPO3 7.6 in Chrome and Firefox.
Tests with TYPO3 8 and Internet Explorer are not done. Maybe somebody else can test this patch in IE and TYPO3 8, especially the hiding of the dragged element in the column.

monosize commented:
I have found a bug. Please wait for an updated PR.

monosize commented:
No bug with this PR. I run into problems with activated xdebug plugin in PHP.

monosize commented:
update grid.css

monosize commented:
Tested in IE with TYPO3 7.6

NamelessCoder commented:
Will do a bit of testing on this when I get home.

NamelessCoder commented:
Overall I like the change. Just two minor things:

The video below shows some jumpiness when using colspan to control the width of columns. Detaching an element causes the container to change widths for the columns.

I'm not completely sure I like the enlarged dropzone effect. Perhaps it would be better if the *actual* drop zone was enlarged on mouse over, so once you enter it, it gets much easier to drop the element correctly (in particular this "hitting the spot" is an issue on my laptop which has a trackpad so a larger drop zone would be neat).

Maybe you have a better idea for that ;)

NamelessCoder commented:
![jumpy](https://cloud.githubusercontent.com/assets/2228052/22480710/a07b9b30-e7f2-11e6-8400-22117143dec2.gif)

NamelessCoder commented:
Oh, forgot to say: the "ctrl to copy" should perhaps not be placed as text but rather some visual change when you hold down ctrl/cmd. Like in the page tree where the drag-n-drop with ctrl shows a different icon.

monosize commented:
A html table is used for the columns and the column width is set by the widest unbreakable content in the cells. I think we must change the css class .flux-grid to
```css
.flux-grid {
    width: 100%;
    table-layout:fixed;
}
```
table-layout:fixed will make the magic ;-)

I will test it.

NamelessCoder commented:
Much better, no more jumpiness 👍

monosize commented:
`\FluidTYPO3\Flux\Backend\TceMain`is extended to handle the `extended paste command` which contains data for the copy via holding the ctrl key during drag and drop.

tested in TYPO3v8.6-dev

monosize commented:
the`extended paste command`also exist in TYPO3 7.6 and the changes in `\FluidTYPO3\Flux\Backend\TceMain`are compatible with TYPO3 7.6

monosize commented:
Work's fine in TYPO3 7.6 ;-) with no text info for holding the ctrl key

monosize commented:
One bug found: during drag and drop with ctrl key.
Copy an element to an empty column: element is copied but flux_parent is not set.
I will check this...

monosize commented:
Drag and drop with copy on first position  in a column now works.
* compatible for flux:grid columns and standard columns
* new data attribute for a flux:grid column 'data-colpos-pastetop=""' to transmit the correct values for the column
  the format is: `PAGEUID-paste-0-CONTENTUID-AREANAME-FLUXCOLPOS`
* in DragDrop.js the data-colpos-pastetop attribute is add to the ajax call.

other changes see comments in the source.

monosize commented:
I forgot: the changes weren't tested in workspaces.

NamelessCoder commented:
I'll give them a run through tomorrow - and I'll also test it with workspaces. Got the proper setup for that running, as you can probably see from my recent commits :)

monosize commented:
Update branch to latest commits https://github.com/FluidTYPO3/flux/commit/281e9dcdbfa430860078843243ec9205263f41e1 and https://github.com/FluidTYPO3/flux/commit/e1fac362884b7a8bf159586556ce758f30d37dac

monosize commented:
Bug Fix: Copy element during drag and drop to first position in flux:column and the column isn't empty.
The sortnumber was wrong and we got every time the default sorting-number `256`.
The call to `$tceMain->getSortNumber` was changed to fetch the lowest sortnumber of the page.

monosize commented:
todo: split grid.css into a 7.6 and 8.6 Versions and use CompatibilityRegistry

monosize commented:
PR is from my side now ready for testing

NamelessCoder commented on Classes/View/PreviewView.php:
```diff
@@ -66,7 +66,7 @@ class PreviewView
                             <div class="t3-grid-cell">
                                 <div class="t3-page-column-header"><div class="t3-page-column-header-label">%s</div></div>
-                                <div data-colpos="%s" class="t3js-sortable t3js-sortable-lang t3js-sortable-lang-%s
+                                <div data-colpos="%s" data-colpos-pastetop="%s" class="t3js-sortable t3js-sortable-lang t3js-sortable-lang-%s
```
The "data-colPos" here should already contain a value that's an integer higher than ContentService::COLPOS_FLUXCONTENT which points to a session-stored array containing the target parent element and column name. That integer-higher-than-ContentService::COLPOS_FLUXCONTENT approach was chosen to avoid all of these customisations and instead use the default "data-colpos" attribute transferring.

monosize commented:
Using now the session-stored array instead of using the attribute data-colpos-pastetop.

NamelessCoder commented:
Hey Frank - I've been testing a bit and it actually looks like we are now compatible enough with the TYPO3 core markup that we can delete our DragAndDrop.js file (which I would much rather do than increase the size of our own override). On LTS a minor CSS fix is needed in grid.css (to prevent max-size of container element when being dragged and when dragging elements into it) but other than that, everything seems to work if I disable our drag-and-drop overrides.

I'll leave some other comments in the code about what is expected to happen in the code you changed - I think maybe the solutions you added, are intended to be handled by other functions and those functions contain bugs. If that's the case we should try to solve those bugs instead of adding new logic.

NamelessCoder commented on Classes/Backend/TceMain.php:
```diff
@@ -149,16 +149,20 @@ public function processCmdmap_beforeStart(&$reference)
      * @return void
      */
-    public function processCmdmap_preProcess(&$command, $table, $id, &$relativeTo, &$reference)
+    public function processCmdmap_preProcess(&$command, $table, $id, &$relativeTo, &$reference, $pasteUpdate)
```
After taking a look into the TYPO3 source code, I think we may be able to move the "postProcess" logic to "preProcess" and skip record updates, instead receiving $pasteUpdate by reference and filling it with the target information. I'll look into that!

NamelessCoder commented on Classes/Backend/TceMain.php:
```diff
@@ -262,6 +268,16 @@ public function processCmdmap_postProcess(&$command, $table, $id, &$relativeTo,
+                    // to avoid merging the colPos after finished the processCmdmap_postProcess
+                    // in the DataHandler
+                    // if colPos < 18181 then the target is outside of a flux:grid
+                    unset ($pasteDataMap[$table][$reference->copyMappingArray[$table][$id]]['colPos']);
```
I'm still not sure we need this... but if we do: setting colPos to COLPOS_FLUXCONTENT (after we've caught and handled the `> COLPOS_FLUXCONTENT` case!) should be the right way to do it.

The change here makes sense as a separate pull request; assuming the problem isn't that our logic in ContentService is doing something wrong when pasting records?

NamelessCoder commented on Classes/Backend/TceMain.php:
```diff
@@ -786,6 +802,15 @@ protected function getClipboardCommand()
+     */
+    protected function getCopyPasteCommand()
+    {
+        $command = GeneralUtility::_GET('cmd');
```
I'm still not convinced we need to read this command. In the method above where this method gets used, the existing case is there to catch a very specific operation: triggering when you do clipboard-paste commands. It must not trigger after drag-n-drop commands (the moveRecord_* methods should take care of that; they must trigger after drag-n-drop ends).

What I'm thinking is that maybe the moveRecord_* method to move-to-top-of-column is doing something wrong or missing some logic? In any case: those are the methods expected to handle this type of command!

NamelessCoder commented on Resources/Public/css/grid.7.6.css:
```diff
@@ -0,0 +1,151 @@
+	margin-right: 10px;
+}
+
+/* copied from TYPO3 v8.6.0 */
```
I would not override these CSS styles on LTS. Flux isn't supposed to homogenise the visual appearance on the two versions - our existing overrides are already vulnerable to changes and we can probably expect more of those! I'd like it if we aim to remove our overrides instead (grid.css is still needed though; and we need an additional patch to prevent height limiting and overflow-hidden on contains you are dragging elements into and which are being dragged).

And that should also mean we don't need the 7.6-specific CSS file - one less override to worry about ;)

NamelessCoder commented on Resources/Public/css/grid.css:
```diff
@@ -1,4 +1,4 @@
-.flux-grid { width: 100%; }
+.flux-grid { width: 100%; table-layout:fixed;}
```
THANK YOU! Seriously, thank you so much for finding this one. I couldn't figure out why my 2/3 column sizing was rendering as 50/50. Excellent, please create this as a separate pull request.  :)

If you do, please add to css as well:

```
.t3-is-dragged .t3-page-ce-body {
	max-height: none;
}
```

That's the hight-limit-overflow-hidden prevention for LTS.

NamelessCoder commented:
Sorry to drag this one out (no pun intended, really). I think we're down to a couple of things that would be great as separate pull requests, I'd rather not merge the other things (the split CSS in particular).

Let's investigate if we really can drop the JS override! That would be a good step imho :)

NamelessCoder commented:
Hi again Frank - since I assume you're sleeping or out having fun, I took the liberty of pushing the CSS table fix and preparing https://github.com/FluidTYPO3/flux/pull/1337 which suggests removing our drag and drop JS override.

monosize commented:
can be closed if #1338 is approved.

NamelessCoder commented:
#1338 merged, closing this issue.
